### ISSUE TITLE
Port forwarding (`multipass forward`)

### ISSUE THREAD

**Issue by Saviq** [canonical/multipass#309]
Forwarding instance ports out onto multipassd's listening address(es):

```
# can pass multiple --forward rules
$ multipass launch --forward <target_port>[:<listen_port>]
Port forwarding configured on:
Port <target_port>: <external_ip>:<listen_port>

$ multipass forward <name> <target_port>[:<listen_port>]
Port forwarding configured on: <external_ip>:<listen_port>

$ multipass unforward <name>[:<target_port>] [<name>[:<target_port>] ...]

$ multipass info <name>
[...]
Port forwarding:
<target_port>  <listen_port>
```

Where it's possible / makes sense, use hypervisor for this, otherwise have multipassd listen.

**Comment by setempler** [comment]
Possible w/ `ssh` local (or remote) port forwarding:

* forward requests from multipass host (port 9000) to multipass instance (port 80)

        sudo ssh \
            -i /var/root/Library/Application\ Support/multipassd/ssh-key/id_rsa \
            -L 9000:localhost:80 \
            multipass@<multipass instance ip>

* access multipass instance (port 80) from any box that can access multipass host

        curl <mutlipass host ip>:9000

Notes:

* ssh key path valid for macOS #913 
* related issues #512 #118

**Comment by JonTheNiceGuy** [comment]
Can I also request another feature?

`bridge <NICNAME>` == `vboxmanage modifyvm "VM name" --nic2 bridged --bridgeadapter2 NICNAME`

`unbridge <NICNAME>` == `vboxmanage modify "VM name" --nic2 none`

**Comment by Saviq** [comment]
@JonTheNiceGuy that would be #118 - remember we need to be able to support the features across all of the hypervisors, not everywhere it is as easy as this :)

**Comment by JonTheNiceGuy** [comment]
> everywhere it is as easy as this :)

Fair comment :)

**Comment by rmetzger** [comment]
The ssh command posted above has slightly changed:
- it's `ssh-keys` instead of `ssh-key`
- the username is `ubuntu` instead of `multipass`.

Resulting in:
```
sudo ssh \
      -i /var/root/Library/Application\ Support/multipassd/ssh-keys/id_rsa \
      -L 9000:localhost:80 \
      ubuntu@<multipass instance ip>
```

**Comment by ktemby** [comment]
In addition to using the SSH tools, I've found it to be very helpful for server use cases to use nginx as well, for the use case of hosting multipass VM's from a headless ubuntu server.

e.g. I'm running a rails webserver in a multipass VM on port 3000, and will link that with ssh onto port 9000 on my host, and be able to connect to my host from anywhere at <hostIP>:8080
     
#### 1. Link multipass to host

    sudo ssh -L 9000:localhost:3000 \ 
        -i /var/snap/multipass/common/data/multipassd/ssh-keys/id_rsa \
        ubuntu@<multipass instance ip> 

#### 2. Enable access to multipass externally from host

Now using nginx, I'm going to connect to port 9000 using the host's IP and expose that on port 8080 for external connections.

    server {
        listen 8080;
        server_name development;
        location / {
            proxy_pass http://localhost:9000;
        }
    }

**Comment by Saviq** [comment]
@ktemby if you have `<multipass instance ip>`, you can give that to nginx directly, no?

**Comment by ktemby** [comment]
@Saviq no, this doesn't work for me.  I think this is the point of multipass port forwarding being a limitation.  While I can hit an IP, I can't tunnel to the port the server runs on without SSH.
e.g. my IP for the instance is 10.151.47.130, the server is running on port 3000 there

    curl 10.151.47.130:3000. 
    curl: (7) Failed to connect to 10.151.47.130 port 3000: Connection refused

**Comment by Saviq** [comment]
> @Saviq no, this doesn't work for me. I think this is the point of multipass port forwarding being a limitation. While I can hit an IP, I can't tunnel to the port the server runs on without SSH.
> e.g. my IP for the instance is 10.151.47.130, the server is running on port 3000 there
> 
> ```
> curl 10.151.47.130:3000
> curl: (7) Failed to connect to 10.151.47.130 port 3000: Connection refused
> ```

More likely your rails webserver only listens on `localhost`, rather than `0.0.0.0` :)

**Comment by ktemby** [comment]
Could be the case - but this is the vanilla rails config.  I also can't curl port 80 from nginx running on the multipass VM either.  As a non-expert, I like it when the tools just work without too much config... which is why I generally have been excited to find multipass!

**Comment by icokk** [comment]
Since `multipass shell` uses ssh anyway, why not create command `multipass forward` which would internally just start ssh with port forwarding? It would be very simple to add, completely cross platform, but it would still be very helpful for users (no need to search for instance IP and multipass ssh key).

**Comment by Saviq** [comment]
@icokk sure, that's the rough approach we're likely to take, we just never got to it yet.

**Comment by mikeubell** [comment]
> sudo ssh \
>       -i /var/root/Library/Application\ Support/multipassd/ssh-keys/id_rsa \
>       -L 9000:localhost:80 \

When I try this it asks for the password.  Even after setting a password for user ubuntu and using that it fails to log in.

**Comment by Saviq** [comment]
@mikeubell it's best to import your own SSH key, no need to reach for the Multipass one. Password SSH is disabled by default. You can enable it manually or set it all up through cloud-init:

https://cloudinit.readthedocs.io/en/latest/topics/modules.html#set-passwords

**Comment by mikeubell** [comment]
@Saviq Thanks.  I found a simpler solution:
ssh -L local_addr:local_port:remote_addr:remote_port -N 127.0.0.1
This does not require a login to the remote multipass instance.

**Comment by Saviq** [comment]
@mikeubell but how does it give you access to the services inside the instance? The line you wrote SSH's into your own host? If you have access to the instance directly, why do you need to do port forwarding?

**Comment by mikeubell** [comment]
I am running a web server in the instance that I need to access from outside the host.  The ssh line forwards the port on that host to the port on the instance.

**Comment by kycfeel** [comment]
Seriously? I don't understand that why everyone else is just talking about the SSH tunneling.

That is just a temporary solution to open the port, for a quick dev or testing environment.

Isn't doing a port forward for VM is a super common use case?

**Comment by Saviq** [comment]
@kycfeel outside of VirtualBox you get direct IP access to the instance. And even with that you can [use `--network`](https://multipass.run/docs/additional-networks) to put the VM on your physical network and gain that access, so port forwarding wasn't a focus, is all.

**Comment by lethargosapatheia** [comment]
Hello. So is there any deadline or approximative time frame when this will be finished?
I was just trying to understand the `--network` option, but the explanation is simply confusing to me. It's not clear at all (except for the bridge part I guess) how I could access the ip of the virtual machine directly:
>   --network <spec>                      Add a network interface to the
                                        instance, where <spec> is in the
                                        "key=value,key=value" format, with the
                                        following keys available:
                                         name: the network to connect to
                                        (required), use the networks command for
                                        a list of possible values, or use
                                        'bridged' to use the interface
                                        configured via `multipass set
                                        local.bridged-network`.
                                         mode: auto|manual (default: auto)
                                         mac: hardware address (default:
                                        random).

**Comment by lethargosapatheia** [comment]
Ok, now I'm starting to get it. There's a local network created just for this, it's not a bridge that allows the VM to connect to your network - that's why I think this is rather ambiguous and it should be made clear in my opinion.

Any way I can assign a static IP to the virtual machine?

**Comment by ricab** [comment]
Hi @lethargosapatheia, when you provide a `--network`, Multipass creates an extra network interface inside that instance and *bridges* it to the specified network (interface) on your host. If you are using `mode=auto` (the default), DHCP is enabled for that additional network interface, inside the instance. It will therefore get an IP as any other machine on the same network (presumably dealt by your router). 

If you want it to have a static IP, you need to configure that yourself on your network (e.g. you could configure your router to always provide the same IP to the right MAC address). You can specify a MAC address in your `--network` parameter, as well as `mode=manual` to configure networking yourself in the instance (e.g. with `netplan`). For example: 

```
$ multipass networks
Name  Type      Description
eth0  ethernet  Ethernet device
[...]
$ multipass launch --network name=eth0,mode=manual,mac=ab:cd:ef:ab:cd:ef`
```

Have a look [here](https://multipass.run/docs/additional-networks) for more info. Hope this helps.

**Comment by lethargosapatheia** [comment]
That would be just ssh tunnelling. I guess the consequence is the same, but it's still not port forwarding in the classical sense.
In any case, it's not that useful, bececause, by default the IP of the virtual machine is accessible, so vscode already knows that IP because you've given it to it. Having a consistent setup on (such as having the same IP each time) is different.

@ricab You mean to say, there's a dhcp server running inside the virtual machine, so inside Ubuntu? How is that called? I thought the dhcp ran somehow on the host's side.

**Comment by ricab** [comment]
> @ricab You mean to say, there's a dhcp server running inside the virtual machine, so inside Ubuntu?

@lethargosapatheia no no, I meant the client side of it. With `mode=auto`, the interface tries to obtain an IP via DHCP. With `mode=manual`, it doesn't.

BTW, if you want to parse, you can use the `--format` option to `list` and `info` to get your output in json, csv, or yaml too.

**Comment by nreith** [comment]
Indeed. It would be reverse tunneling with socks proxy. NAT would give similar behavior in passing traffic through host to take advantage of network access. But a nice feature of vscode dynamic port forwarding and even wsl2 is WebApps appear on localhost. No need to edit url for the current IP of multipass vm to get it to work. I came across this thread because I was looking for ways to make multipass act a little more like wsl2. I'd like to use this with cloud-init to give a day zero dev box to software engineers or data scientists who are scared of Linux.

**Comment by yingshaoxo** [comment]
You don't need port forwarding

VScode can do it for you if you login by using vscode-remote:

https://github.com/yingshaoxo/ubuntu_ssh_server_for_the_stupid_apple_macos_m1_system#another-method-use-multipass

**Comment by NickFranceschina** [comment]
been monkeying with this stuff for days and the way @ricab explained it made more sense to me than the docs... so I used `--network en1` for my 2020 Mac M1 OSX and then I got an actual IP address from my actual router and I could finally reach my containers directly.   I just wanted to point out that from my home router's perspective, all the VMs seem to have the same MAC address.. the MAC of the host.. which isn't "wrong", but it was at first confusing me and my cheap home router I guess (the "address reservation" tooling in my home TPLink router assumes every MAC it sees should have only one IP).  I built several more VMs where I passed in a MAC address to use, and when I shell into them I do indeed see that it's using the supplied MAC address, but I guess that's not the same thing since the VMs are living on their own internal network of sorts.   Just wanted to leave this note for anyone that might be googling these search terms.  This networking stuff is maddening and well above my pay grade :)

**Comment by ricab** [comment]
Thank you @NickFranceschina, that's interesting, I also have just a basic internet-provider router, but it reports the instance's MAC address (even knows its name). That's with a bridge created by Multipass via NetworkManager...

**Comment by ricab** [comment]
Hi @NickFranceschina, I guess you're using the `qemu` driver (`multipass get local.driver`) on macOS, correct? We use [`vmnet`](https://developer.apple.com/documentation/vmnet?language=objc) in bridged mode, which is very poorly documented and essentially a black box to us what exactly Apple does. On top of that, routers can report DHCP leases in different ways and there can be inconsistencies at that level too.

I tried what you described and, when passing a wifi interface to `--network` my router doesn't even list the device! If I connect an ethernet cable over USB (no ethernet port on the macbook) and use that interface, then I do see the instance with the correct MAC address in the router's list. One other thing you could try is to [create a virtual bridge](https://support.apple.com/guide/mac-help/bridge-virtual-network-interfaces-on-mac-mh43557/mac) manually and use that. I am afraid I don't have much more than that...

**Comment by Saviq** [comment]
Let me just add that bridging over WiFi is generally problematic, because of how packets are routed over the wireless link. Simplifying, the access point doesn't expect packets from a different MAC on the link (there are no WiFi switches like there are for Ethernet).

The fact that it works at all suggests macOS is doing some translation (think masquerade) under the hood - but that is going to be hit and miss on the other side.

Wonder if the IPs are actually/also sitting on the WiFi interface in some way…

**Comment by NickFranceschina** [comment]
Thanks guys... @Saviq  yes I should actually have this box plugged into the ethernet instead, but at the moment I only have one port in the router (already used) and no extra switches laying around, so I tried with the wifi.  @ricab I'm trying to do everything as "vanilla" as possible with multipass, falling back to the defaults that I'm sure have been thoroughly considered (so yeah I'm using the qemu driver).  I don't want to dig into any specialty settings on OSX networking stuff as I tried that before and ended up in a corner where I had to re-install the OS to be sane again. 

So when I get a chance I will try this again with the ethernet adapter and see if the mac addresses work as expected.  until then I've opted to configure static IPs (which makes sense anyways since these services don't need DHCP) and it all appears to be working for now.  I thank you for your feedback!

**Comment by NickFranceschina** [comment]
@ricab yeah so again I haven't spent much of my life in Linux so I'm doing things as best I can figure out... I originally created the VM with just `--network en1` (I didn't specify "manual" so it did auto).... then I created a `sudo nano /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg` file with the `network: {config: disabled}` (per the notes at the top of `/etc/netplan/50-cloud-init.yaml`) and then created a new `/etc/netplan/01-netcfg.yaml` with the interface name and the static ip/gateway/nameservers I wanted.... then `sudo netplan apply`

I think that's what I did anyways :)   I should def try to break it all and see if I can fix it to make sure I really understand

**Comment by NickFranceschina** [comment]
@Saviq @ricab --- an update:  added a switch and changed everything over to use ethernet DHCP instead of wifi and now I'm seeing different mac addresses from the router for the VMs!  nice.  thanks for your insights!

**Comment by ricab** [comment]
You're welcome, thanks for sharing your experience!

**Comment by zippy-zebu** [comment]
How does it supposed to work with exposing instances with MULTIPASS_SERVER_ADDRESS parameter ? Described [here](https://multipass.run/docs/how-to-use-multipass-remotely-a-preview).

**Comment by akauppi** [comment]
While I'd still like seeing port-forwarding as a Multipass feature, some year, it can be pretty well worked around. The posts (from 2019, 2020) above already gave the grunt of the commands, but I've added a bit extra. Perhaps this helps someone.

https://github.com/akauppi/mp/blob/main/npm/.mp.dive.sh

The idea is that I'll copy-paste that script to any repo where I use Multipass, as a developer. Note: I lay out repos so that Multipass is not a dependency (native development is possible), but an option for more sandboxing and control of OS level build dependencies.

Compared with the above approaches, this:

- runs as a normal user, no host side `sudo` is needed. I ask myself to copy the Multipass private key to a destination within the user account.
- can change to a working directory, with the same `shell`-like command

```
ssh -i ${_LOCAL_KEY} -t -q ubuntu@${_MP_IP} ${_PORT_PARAMS} "cd /home/ubuntu/${_FOLDER} && exec bash -i"
```

Just my toolchain.
